File: python/src/siphon/providers/custom.py
"""Generic streaming multipart uploader for custom/self-hosted providers."""

from __future__ import annotations

import secrets
from typing import AsyncIterator, Iterator

import httpx

from siphon.providers import ProviderRef


def multipart_chunks(
    field_name: str,
    filename: str,
    mime_type: str,
    chunks: Iterator[bytes],
    boundary: str,
) -> Iterator[bytes]:
    for label, value in (
        ("field_name", field_name),
        ("filename", filename),
        ("mime_type", mime_type),
    ):
        if "\r" in value or "\n" in value:
            raise ValueError(
                f"{label} must not contain CR or LF characters (got {value!r}); "
                "this would allow injecting extra multipart headers or fields."
            )
    preamble = (
        f"--{boundary}\r\n"
        f'Content-Disposition: form-data; name="{field_name}"; filename="{filename}"\r\n'
        f"Content-Type: {mime_type}\r\n\r\n"
    ).encode("utf-8")
    yield preamble
    for chunk in chunks:
        yield chunk
    yield f"\r\n--{boundary}--\r\n".encode("utf-8")


class GenericHTTPUploadProvider:
    """A minimal Provider for a self-hosted endpoint that accepts multipart uploads
    and returns JSON with an "id" field. Subclass or wrap this to match a specific
    endpoint's request/response shape.
    """

    def __init__(
        self,
        upload_url: str,
        field_name: str = "file",
        http_client: httpx.Client | None = None,
    ) -> None:
        self._upload_url = upload_url
        self._field_name = field_name
        self._client = http_client or httpx.Client()

    def supports_media_type(self, mime_type: str) -> bool:
        return True

    def inline_size_limit(self) -> int:
        return 0  # no fixed inline convention for an arbitrary custom endpoint

    def upload(self, chunks: Iterator[bytes], mime_type: str) -> ProviderRef:
        boundary = f"SiphonBoundary{secrets.token_hex(16)}"
        body = multipart_chunks(self._field_name, "upload", mime_type, chunks, boundary)
        response = self._client.post(
            self._upload_url,
            content=body,
            headers={"Content-Type": f"multipart/form-data; boundary={boundary}"},
        )
        response.raise_for_status()
        data = response.json()
        return ProviderRef(id=data["id"], expires_at=data.get("expires_at"))

    async def upload_async(
        self, chunks: AsyncIterator[bytes] | Iterator[bytes], mime_type: str
    ) -> ProviderRef:
        if hasattr(chunks, "__anext__"):
            collected = [c async for c in chunks]  # type: ignore[union-attr]
        else:
            collected = list(chunks)  # type: ignore[arg-type]
        boundary = f"SiphonBoundary{secrets.token_hex(16)}"
        body = b"".join(multipart_chunks(self._field_name, "upload", mime_type, iter(collected), boundary))
        async with httpx.AsyncClient() as client:
            response = await client.post(
                self._upload_url,
                content=body,
                headers={"Content-Type": f"multipart/form-data; boundary={boundary}"},
            )
        response.raise_for_status()
        data = response.json()
        return ProviderRef(id=data["id"], expires_at=data.get("expires_at"))

    def build_reference_block(self, ref: ProviderRef, mime_type: str) -> dict:
        return {"type": "file_reference", "id": ref.id}

    def build_inline_block(self, base64_data: str, mime_type: str) -> dict:
        return {"type": "inline_base64", "mime_type": mime_type, "data": base64_data}

File: typescript/src/providers/custom.ts
import { randomBytes } from "node:crypto";
import type { Provider, ProviderRef } from "./types.js";

export async function* multipartChunks(
  fieldName: string,
  filename: string,
  mimeType: string,
  chunks: AsyncIterable<Buffer>,
  boundary: string
): AsyncIterable<Buffer> {
  for (const [label, value] of [
    ["fieldName", fieldName],
    ["filename", filename],
    ["mimeType", mimeType],
  ] as const) {
    if (value.includes("\r") || value.includes("\n")) {
      throw new Error(
        `${label} must not contain CR or LF characters (got ${JSON.stringify(value)}); ` +
          "this would allow injecting extra multipart headers or fields."
      );
    }
  }
  const preamble =
    `--${boundary}\r\n` +
    `Content-Disposition: form-data; name="${fieldName}"; filename="${filename}"\r\n` +
    `Content-Type: ${mimeType}\r\n\r\n`;
  yield Buffer.from(preamble, "utf-8");
  for await (const chunk of chunks) {
    yield chunk;
  }
  yield Buffer.from(`\r\n--${boundary}--\r\n`, "utf-8");
}

function toWebReadableStream(chunks: AsyncIterable<Buffer>): ReadableStream<Uint8Array> {
  const iterator = chunks[Symbol.asyncIterator]();
  return new ReadableStream<Uint8Array>({
    async pull(controller) {
      const { value, done } = await iterator.next();
      if (done) {
        controller.close();
      } else {
        controller.enqueue(value);
      }
    },
  });
}

export class GenericHTTPUploadProvider implements Provider {
  constructor(private uploadUrl: string, private fieldName: string = "file") {}

  supportsMediaType(mimeType: string): boolean {
    return true;
  }

  inlineSizeLimit(): number {
    // Sentinel: 0 (or any non-positive value) means "no declared inline size
    // limit" for this provider, since an arbitrary custom HTTP endpoint has no
    // fixed inline convention to report. This is NOT "zero bytes allowed" --
    // the orchestrator treats any inlineSizeLimit() <= 0 as "unbounded" and
    // skips the inline-size check entirely. Providers with a genuine inline
    // size cap should return a positive byte count instead.
    return 0;
  }

  async upload(chunks: AsyncIterable<Buffer>, mimeType: string): Promise<ProviderRef> {
    const boundary = `SiphonBoundary${randomBytes(16).toString("hex")}`;
    const body = toWebReadableStream(multipartChunks(this.fieldName, "upload", mimeType, chunks, boundary));

    const response = await fetch(this.uploadUrl, {
      method: "POST",
      headers: { "Content-Type": `multipart/form-data; boundary=${boundary}` },
      body,
      duplex: "half",
    } as RequestInit & { duplex: "half" });

    if (!response.ok) {
      throw new Error(`Upload to ${this.uploadUrl} failed with status ${response.status}`);
    }
    const data = (await response.json()) as { id: string; expires_at?: number };
    return { id: data.id, expiresAt: data.expires_at ?? null };
  }

  buildReferenceBlock(ref: ProviderRef, mimeType: string): Record<string, unknown> {
    return { type: "file_reference", id: ref.id };
  }

  buildInlineBlock(base64Data: string, mimeType: string): Record<string, unknown> {
    return { type: "inline_base64", mime_type: mimeType, data: base64Data };
  }
}
